### PR TITLE
fix: don't error if source selector selects no sources

### PR DESCRIPTION
**Pull request by erikgb**
Fixes https://github.com/cert-manager/trust-manager/issues/426?

~~We have a test that asserts an error when the bundle is empty. I doubt if this is correct, and to fix https://github.com/cert-manager/trust-manager/issues/426 properly, I think this behavior must go. We could eventually assert that a Bundle is invalid if it doesn't contain at least one source spec. But that is better done with the webhook (or API CEL expression) IMO.~~

UPDATE: After discussing this PR in today's stand-up, we decided to keep the existing (and tested) behavior where a target bundle needs to contain at least one certificate, at least for now. This means that this PR only fixes a minor bug: before this PR it would error if any source selector selects no sources. Now this is ok, but the resulting bundle must contain at least one certificate.

### CODE REVIEW

**Comment by erikgb** [comment]
@SgtCoDFish I think this PR is ready for another review now!

File: pkg/bundle/source.go [hunk 1]
```diff
@@ -81,6 +81,12 @@ func (b *bundle) buildSourceBundle(ctx context.Context, sources []trustapi.Bundl
 			return bundleData{}, fmt.Errorf("failed to retrieve bundle from source: %w", err)
 		}
 
+		// A source selector may select no configmaps/secrets, and this is not an error.
+		// Must continue early since CertPool does not currently support adding no certificates.
+		if len(sourceData) == 0 {
+			continue
+		}
```
Comment (SgtCoDFish) [inline pkg/bundle/source.go:88]: suggestion: Would it be worth logging a warning in this situation? As a platform admin, I'd want to either remove a selector which selects nothing, or fix it.

**Review by SgtCoDFish — COMMENTED** [review]
One suggestion from me; what do you think?

File: pkg/bundle/source.go (hunk 1 above)
Comment (erikgb) [reply pkg/bundle/source.go:88]: I agree that a platform admin needs to understand what the operator is doing (and not doing). It seems like events are used for similar scenarios in this project. Should we use an event for this too, or just log it? With contextual logging, I am always in doubt what you exactly mean with a "logging a warning". 😉

Comment (SgtCoDFish) [reply pkg/bundle/source.go:88]: I'd personally have just logged it, and wouldn't have thought about events. I don't think an event would be bad, but I think the lightweight option of logging is totally fine!

Comment (erikgb) [reply pkg/bundle/source.go]: I'm now logging if a label selector matches no resources. Not having log levels feels a bit strange to me, but supported by Dave Chaney, I am proposing to log this on a normal info level.

> Eliminate the warning level, it’s either an informational message, or an error condition.

Ref. https://dave.cheney.net/tag/logging

I also introduced a custom error to eliminate the risk of other errors slipping through the empty string conditional. PTAL!

File: pkg/bundle/source.go [hunk 2]
```diff
@@ -171,10 +180,10 @@ func (b *bundle) secretBundle(ctx context.Context, ref *trustapi.SourceObjectKey
 		if selectorErr != nil {
 			return "", fmt.Errorf("failed to parse label selector as Selector for Secret in namespace %s: %w", b.Namespace, selectorErr)
 		}
-		if err := b.client.List(ctx, &sl, client.MatchingLabelsSelector{Selector: selector}); apierrors.IsNotFound(err) {
```
Comment (erikgb) [inline pkg/bundle/source.go:174]: A list call will only return 404 if something is really wrong, ref. https://github.com/kubernetes/client-go/issues/537#issuecomment-450030586

Comment (SgtCoDFish) [reply pkg/bundle/source.go:174]: Awesome context, thank you!

**Comment by erikgb** [comment]
> I'm not sure if I missed something but wasn't there a new check to ensure the bundle wasn't empty, so we still required a bundle to resolve to at least one cert?
> 
> Is that missing here? Was it added elsewhere? I might be missing it but it looks like this would permit an empty target?

The check is already there: https://github.com/cert-manager/trust-manager/blob/main/pkg/bundle/source.go#L90-L92, and it is now not changed in this PR. 😉

**Review by SgtCoDFish — APPROVED** [review]
/lgtm
/approve

Sorry for the delay, something came up!

I totally missed that check. Thanks very much for this 🙌

**Pull request merged**